flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -110,22 +112,43 @@ class AbstractArrayBox {
     return lbounds;
   }
 
+  mlir::Value getSlice() const { return slice; }
+
   bool lboundsAllOne() const { return lbounds.empty(); }
+  bool isContiguous() const { return !slice; }
+  // We are using slice to handle non contiguous input
+  // descriptor. But in this case we need to apply lbounds
+  // twice (once on the slice indexes, and another time on
+  // coordinate array references).
+  bool applyLboundsBeforeSlice() const { return !sliceIsOneBased; }
 
 protected:
+  // TODO: replace extents and lbounds by:
+  // mlir::Value shape; \\ fir::ShapeType or fir::ShapeShiftType
   llvm::SmallVector<mlir::Value, 4> extents;
   llvm::SmallVector<mlir::Value, 4> lbounds;
+  // If this is not a contiguous array, indicate the slice
+  mlir::Value slice;
+  // If there is a slice, does the ArrayBox is one based or do
+  // lbounds still need to be considered ? This DOES NOT has anything to do with
+  // whether the slice lower bounds are one or not.
+  bool sliceIsOneBased;
```

Comment:
The all mess from sliceIsOneBased comes from the fact that this proposal is using SliceOp to represent non contiguous assumed shape arguments. These Boolean should really be called `sliceIsFakeSliceToRepresentDescriptorStride`.
```
subroutine foo(x)
  real :: x(40:)
  print *, x(42)
end subroutine
```
With this PR, we represent x as a slice from a contiguous array y whose address is the one from the box `x = y(40: box_extent: box_stride/box_elem_size)`.

Which means we handle `x(42)` as if it were `y(42:n:step)(42)` (this last expression is not a Fortran thing, a user cannot index a stride).  If we lower this to `fir.array_op %shape_shift %slice %42`., then code-gen will actually index `x(83)` because it expects that if there is a slice, the lower bounds from the shape_shift applies to the slice bounds, but not to the coordinates arguments.
This makes sense because  `y(42:n:step)(42)` is not a Fortran thing, so when there is an array_op with a slice, the coordinates should usually be compiler generated indexes looping over the extents.

Anyway, this kind of weirdness that does not quite fit is exactly why I am not a fan of using a slice to represent the discontinuity of an assumed shape.